src/classes/EP_ManageEPTemplate_CTRL.cls

```diff
@@ -247,7 +257,30 @@ public with sharing class EP_ManageEPTemplate_CTRL {
             source.setRedirect(true);
             source.getParameters().put('t',''+(System.currentTimeMillis()));
             return source;
+
         } catch (Exception ex) {
+            Database.rollback(sp);
+
+            //rollback new Template creation
+            template.Id = templateId;
+
+            //rollback Task insert
+            for (Engagement_Plan_Task__c task : tasksForInsert) {
+                task.Id = null;
+            }
+            //rollback Task Parent assignment
+            for (TaskWrapper wrapper : mapTaskWrappers.values()) {
+                Engagement_Plan_Task__c epTask = wrapper.detail;
+
+                Id existingParentId = epTask.Id == null
+                    ? null
+                    : existingParentIdsByTaskId.get(epTask.Id);
+
+                epTask.Parent_Task__c = existingParentId != null
+                    ? existingParentId
+                    : wrapper.parent != null ? wrapper.parent.detail.Id : null;
+            }
+
```

Comment:
interesting!  so even though we've done the rollback, we have to manually fixup the state of any data structures we may have modified during the save.  I don't believe we discussed that during our team discussions about transaction handling.  It makes sense when you think about it.  Could you revisit your transaction investigation document and make sure this is clearly mentioned in it.  thanks!

Reply:
When a user fixes a cause for the error and clicks Save again, an error might be generated due modifications applied in the first Save (insert of sObject will not clear the Id on the rollback for example). However, not every method having Database.Rollback() applied needs to "rollback" its variables.
The transaction document has been updated to include the reset of sObjects when rollback is applied. Please let me know if anything more should be added to the document.

Reply:
doc edits looks good.  thanks.